tests/unit_tests.cpp

```diff
@@ -326,6 +326,75 @@ BOOST_AUTO_TEST_CASE(buffer_versioning) {
   }
 }
 
+// TODO: Extend this
+BOOST_AUTO_TEST_CASE(buffer_api) {
+  namespace s = cl::sycl;
+
+  s::buffer<int, 1> buf_a(32);
+  s::buffer<int, 1> buf_b(32);
+  auto buf_c = buf_a;
+
+  BOOST_REQUIRE(buf_a == buf_a);
+  BOOST_REQUIRE(buf_a != buf_b);
+  BOOST_REQUIRE(buf_a == buf_c);
+}
+
+// TODO: Extend this
+BOOST_AUTO_TEST_CASE(accessor_api) {
+  namespace s = cl::sycl;
+
+  s::buffer<int, 1> buf_a(32);
+  s::buffer<int, 1> buf_b(32);
+  auto buf_c = buf_a;
+
+  const auto run_test = [&](auto get_access) {
+    auto acc_a1 = get_access(buf_a);
+    auto acc_a2 = acc_a1;
+    auto acc_a3 = get_access(buf_a, s::range<1>(16));
+    auto acc_a4 = get_access(buf_a, s::range<1>(16), s::id<1>(4));
+    auto acc_a5 = get_access(buf_a);
+    auto acc_b1 = get_access(buf_b);
+    auto acc_c1 = get_access(buf_c);
+
+    BOOST_REQUIRE(acc_a1 == acc_a1);
+    BOOST_REQUIRE(acc_a2 == acc_a2);
+    BOOST_REQUIRE(acc_a1 != acc_a3);
+    BOOST_REQUIRE(acc_a1 != acc_a4);
+    BOOST_REQUIRE(acc_a1 != acc_b1);
+    // NOTE: A strict reading of the 1.2.1 Rev 7 spec, section 4.3.2 would imply
+    // that these should not be equal, as they are not copies of acc_a1.
+    BOOST_REQUIRE(acc_a1 == acc_a5);
+    BOOST_REQUIRE(acc_a1 == acc_c1);
```

Comment:
When merging this into the new runtime, I realized that these two tests cause trouble here. With the new runtime, an accessor does the following:
1. On construction, binds itself to a buffer by storing a `shared_ptr` on the host side to the `rt::data_region` (which is basically what `buffer_impl` used to be)
2. When binding to a `handler` object, stores a pointer to the `rt::requirement` handle that uniquely describes the requirement in the DAG
3. On kernel submission, this pointer is exchanged for the actual device pointer, after the scheduler has decided on which device the operation should run.

The problem is that these tests are run in command group scope (not kernel scope), so between steps 2 and 3. In that moment, device accessors for the same buffer and access range are uniquely identifiable because of the `rt::requirement` pointer that they hold. After step 3, they are again indistinguishable again.
Host accessors are different however: Since constructing a host accessor is a synchronous operation, after construction they are already after step 3) and hold a data pointer. They are hence indistinguishable from other host accessors to the same data region.

This means that for the device accessor tests, the compared accessors will be different while for host accessors they are not.

Is comparison of device accessors in command group scope important to you? This might be a grey area of the spec, as using device accessors on the host is illegal, and it is not entirely clear if this might already be a form of "use".
If this is not critical for you, I would propose to remove those tests for now - the proper fix imo would be to compare device accessors in a kernel, which could be implemented in the tests in a subsequent PR.

Reply:
No, it's not at all important to me! I've chosen to do the comparison within the CGF because it was easier ;-). While reading up on accessors again for this PR I've also noticed that the whole concept of "use" is very vague, and I've been thinking about opening a corresponding issue in the spec repository.

Reply:
Thanks, I will then comment out these two tests for now in the branch for the new runtime.

> I've been thinking about opening a corresponding issue in the spec repository.

Please do :) It's not the first time this vagueness of the spec has caused some issue, e.g. [here](https://stackoverflow.com/questions/61741313/sycl-dpc-accessor-vs-global-ptr-in-kernel-function-object) it was attempted to use `accessor::get_pointer()` in command group scope, which also caused trouble..